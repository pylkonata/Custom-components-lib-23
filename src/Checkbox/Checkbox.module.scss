* {
  box-sizing: border-box;
  outline: 0;
  border: 0;
  margin: 0;
}

.container {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background-color: transparent;
  cursor: pointer;

  &.disabled {
    pointer-events: none;
    cursor: default;
  }
}

.checkbox {
  width: 2.625rem;
  height: 2.625rem;
  padding: 0.5625rem;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  transition: background-color 0.2s cubic-bezier(0.4, 0, 0.2, 1) 0ms;


  &:hover {
    border-radius: 50%;
    background-color: #1976d20a;
    transition: background-color 0.2s cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  }

  .input {
    width: 1.125rem;
    height: 1.125rem;
    cursor: pointer;
  }
}

.label {
  cursor: pointer;
}
