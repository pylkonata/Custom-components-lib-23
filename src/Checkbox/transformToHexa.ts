import { CSSProperties } from 'react';

export const transformToHexa = (hex: CSSProperties['color'], alpha: number) => {
	const color = `${hex}${Math.floor(alpha * 255)
		.toString(16)
		.padStart(2, '0')}`;
	return color;
};
