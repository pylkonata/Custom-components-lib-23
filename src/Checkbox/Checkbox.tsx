import { CSSProperties, ChangeEvent, useEffect, useId, useRef, useState, MouseEvent } from 'react';
import s from './Checkbox.module.scss';

import { transformToHexa } from './transformToHexa';

export interface CheckboxProps {
	checked?: boolean;
	label?: string;
	color?: CSSProperties['color'];
	disabled?: boolean;
	onChange?: () => void;
}

const Checkbox = ({ checked = false, label, color, disabled = false, onChange }: CheckboxProps) => {
	const id = useId();
	const [isChecked, setIsChecked] = useState(checked);
	const [onHover, setOnHover] = useState(false);
	const colorBackRef = useRef('');

	useEffect(() => {
		if (color) {
			colorBackRef.current = transformToHexa(color, 0.04);
		}
	}, [color]);

	const containerClasses = [s['container'], disabled ? s['disabled'] : ''].join(' ');

	const labelClasses = [s['label'], disabled ? s['disabled'] : ''].join(' ');

	const checkboxStyle = {
		backgroundColor: onHover && color ? colorBackRef.current : '',
	};

	const onChangeCheckbox = (
		event: ChangeEvent<HTMLInputElement> | MouseEvent<HTMLDivElement, globalThis.MouseEvent>
	) => {
		if (event.currentTarget !== event.target) {
			return;
		}

		setIsChecked((prev) => !prev);
		if (onChange) {
			onChange();
		}
	};
	const handleOnHover = () => {
		if (color && !disabled) {
			setOnHover((prev) => !prev);
		}
	};

	return (
		<div className={containerClasses} data-testid='container'>
			<div
				className={s['checkbox']}
				onMouseEnter={handleOnHover}
				onMouseLeave={handleOnHover}
				style={checkboxStyle}
				onClick={onChangeCheckbox}
				data-testid='checkboxWrap'
			>
				<input
					type='checkbox'
					id={`input-${id}`}
					style={{ accentColor: color ? color : '' }}
					onChange={onChangeCheckbox}
					checked={isChecked}
					disabled={disabled}
					className={s['input']}
				/>
			</div>
			{label && (
				<label htmlFor={`input-${id}`} className={labelClasses} data-testid='label'>
					{label}
				</label>
			)}
		</div>
	);
};

export default Checkbox;
