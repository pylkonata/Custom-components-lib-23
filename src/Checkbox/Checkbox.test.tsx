import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Checkbox from './Checkbox';

describe('Checkbox', () => {
	test('render basic Checkbox', () => {
		render(<Checkbox />);
		const checkbox = screen.getByRole('checkbox');
		const label = screen.queryByTestId('label');

		expect(checkbox).toBeInTheDocument();
		expect(label).not.toBeInTheDocument();
		expect(checkbox).toBeEnabled();
		expect(checkbox).toMatchSnapshot();
	});

	test('render Checkbox with label and color', () => {
		render(<Checkbox label='Label text' color='#2e7d32' />);
		const checkbox = screen.getByRole('checkbox');
		const checkboxWrap = screen.getByTestId('checkboxWrap');
		const label = screen.getByTestId('label');

		expect(checkbox).toBeInTheDocument();
		expect(label).toBeInTheDocument();
		expect(label).toHaveTextContent('Label text');
		fireEvent.mouseEnter(checkboxWrap);
		expect(screen.getByTestId('checkboxWrap')).toHaveStyle({
			backgroundColor: '#2e7d320a',
		});
	});

	test('render checked and disabled Checkbox', async () => {
		const user = userEvent.setup();
		const handleOnChange = jest.fn();
		render(<Checkbox checked={true} disabled={true} onChange={handleOnChange} />);

		const checkbox = screen.getByRole('checkbox');
		const container = screen.getByTestId('container');

		expect(checkbox).toBeDisabled();
		expect(checkbox).toBeChecked();

		expect(container).toHaveClass('container', 'disabled');
		await user.click(checkbox);
		expect(handleOnChange).not.toHaveBeenCalled();
	});

	test('render Checkbox with onChange props', async () => {
		const user = userEvent.setup();
		const handleOnChange = jest.fn();
		render(<Checkbox checked={false} onChange={handleOnChange} />);

		const checkbox = screen.getByRole('checkbox');

		expect(checkbox).not.toBeChecked();
		await user.click(checkbox);
		expect(screen.getByRole('checkbox')).toBeChecked();
		expect(handleOnChange).toHaveBeenCalledTimes(1);
	});
});
