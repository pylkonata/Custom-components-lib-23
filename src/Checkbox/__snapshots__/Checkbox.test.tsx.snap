// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox render basic Checkbox 1`] = `
<input
  class="input"
  id="input-:r0:"
  type="checkbox"
/>
`;
